daffodil-runtime1-unparser/src/main/scala/org/apache/daffodil/processors/unparsers/ExpressionEvaluatingUnparsers.scala

```diff
@@ -77,30 +78,83 @@ final class SetVariableUnparser(
 
 }
 
+final class NewVariableInstanceSuspendableExpression(
+  override val expr: CompiledExpression[AnyRef],
+  override val rd: VariableRuntimeData)
+  extends SuspendableExpression {
+
+  override protected def processExpressionResult(ustate: UState, v: DataValuePrimitive): Unit = {
+    val vi = ustate.variableMap.find(rd.globalQName)
+    Assert.invariant(vi.isDefined)
+    // Reset original variable's state from VariableInProcess
+    vi.get.setState(vi.get.priorState)
+    ustate.variableMap.newVariableInstance(rd, v)
+  }
+
+  override protected def maybeKnownLengthInBits(ustate: UState) = MaybeULong(0)
+}
+
 // When implemented this almost certainly wants to be a combinator
 // Not two separate unparsers.
-class NewVariableInstanceStartUnparser(override val context: RuntimeData)
+class NewVariableInstanceStartUnparser(override val context: VariableRuntimeData)
   extends PrimUnparserNoData {
 
   override lazy val runtimeDependencies = Vector()
 
   override lazy val childProcessors = Vector()
 
   override def unparse(state: UState) = {
-    val vrd = context.asInstanceOf[VariableRuntimeData]
-    state.newVariableInstance(vrd)
+    if (context.maybeDefaultValueExpr.isDefined) {
+      val maybeVar = state.variableMap.find(context.globalQName)
+      Assert.invariant(maybeVar.isDefined)
+
+      // Set the variable's state to VariableInProcess while waiting for the NVI
+      // expression to complete. When the expression completes, this original
+      // variable instance will have its state reset to its original value.
+      maybeVar.get.setState(VariableInProcess)
+      val suspendableExpression = new NewVariableInstanceSuspendableExpression(context.maybeDefaultValueExpr.get, context)
+      suspendableExpression.run(state)
```

Comment:
I'm not sure if this is right. The NewVariableInstanceSuspendableExpression doesn't create the actual newVariableInstance until the deafult expression successfully evaluates. Isn't it possible that this default requires a suspension, then we continue and then an expression tries to reference that new instance. But since it's not created yet it's still trying to access the old instance?

Seems like this should be something like
```scala
state.variableMap.newVariableInstance(context)
val maybeVar = state.variableMap.find(...) // get the new instsance
maybeVar.get.setState(VariableInProcess)
```
So we create a new instance, and then set its state so that anything referencing this new instance needs to wait until the default is evaluated. Then the NVISuspendableExpression processExpressionResult function sets the value of this new instance and sets the value to VariableSet. Or something along those lines?

Reply:
So I had been trying to get the approach you described to work for quite a while, but kept running into SuspensionDeadlock issues.  Perhaps I wasn't copying things appropriately when the UState gets cloned for suspension.  I'll take another look at it, but this commit was the first thing that I could get to work and handle scoping properly.